ir/reader.go

```diff
@@ -122,26 +132,72 @@ func (self *IrReader) ReadToFunc(r io.Reader, f func(ffi.LogEvent) bool) (ffi.Lo
+	}
+	return self.ReadToFunc(fn)
+}
+
+// Read the CLP IR stream until [strings/HasPrefix] returns true for a
+// [ffi.LogEventView] and the given prefix. Errors are propagated from
+// [ReadToFunc].
+func (self *Reader) ReadToPrefix(prefix string) (*ffi.LogEventView, error) {
+	fn := func(event *ffi.LogEventView) bool {
+		return strings.HasPrefix(event.LogMessageView, prefix)
+	}
+	return self.ReadToFunc(fn)
 }
 
-// Read the IR stream using the io.Reader until [bytes/Contains] returns true
-// for [ffi.LogEvent.Msg] and subslice. Errors are propagated from ReadNextLogEvent.
-func (self *IrReader) ReadToContains(r io.Reader, subslice []byte) (ffi.LogEvent, error) {
-	return self.ReadToFunc(r, func(e ffi.LogEvent) bool { return bytes.Contains(e.Msg, subslice) })
+// Read the CLP IR stream until [strings/HasSuffix] returns true for a
+// [ffi.LogEventView] and the given suffix. Errors are propagated from
+// [ReadToFunc].
+func (self *Reader) ReadToSuffix(suffix string) (*ffi.LogEventView, error) {
+	fn := func(event *ffi.LogEventView) bool {
+		return strings.HasSuffix(event.LogMessageView, suffix)
+	}
+	return self.ReadToFunc(fn)
 }
 
-// Read the IR stream using the io.Reader until [bytes/HasPrefix] returns true
-// for [ffi.LogEvent.Msg] and prefix. Errors are propagated from ReadNextLogEvent.
-func (self *IrReader) ReadToPrefix(r io.Reader, prefix []byte) (ffi.LogEvent, error) {
-	return self.ReadToFunc(r, func(e ffi.LogEvent) bool { return bytes.HasPrefix(e.Msg, prefix) })
+// fillBuf shifts the remaining valid IR in [Reader.buf] to the front and then
+// calls [ioReader.Read] to fill the remainder with more IR. Before reading into
+// the buffer, it is grown by 1.5x if more than 1/4th of it is unconsumed IR.
+// Forwards the return of [ioReader.Read].
+func (self *Reader) fillBuf() (int, error) {
+	if (self.end - self.start) > len(self.buf)>>2 {
+		buf := make([]byte, len(self.buf)+len(self.buf)/2)
+		copy(buf, self.buf[self.start:self.end])
+		self.buf = buf
+	} else {
+		copy(self.buf, self.buf[self.start:self.end])
+	}
+	self.end -= self.start
+	self.start = 0
+	n, err := self.read()
+	return n, err
 }
 
-// Read the IR stream using the io.Reader until [bytes/HasSuffix] returns true
-// for [ffi.LogEvent.Msg] field and suffix. Errors are propagated from ReadNextLogEvent.
-func (self *IrReader) ReadToSuffix(r io.Reader, suffix []byte) (ffi.LogEvent, error) {
-	return self.ReadToFunc(r, func(e ffi.LogEvent) bool { return bytes.HasSuffix(e.Msg, suffix) })
+// read is a wrapper around a io.Reader.Read call. It uses the correct range in
+// buf and adjusts the range accordingly. Always returns the number of bytes
+// read. On success nil is returned. On failure an error is forwarded from
+// [io.Reader], unless n > 0 and io.EOF == err as we have not yet consumed the
+// CLP IR.
+func (self *Reader) read() (int, error) {
+	n, err := self.ioReader.Read(self.buf[self.end:])
+	self.end += n
+	if nil != err && io.EOF != err {
+		return n, err
+	}
```

Comment:
```suggestion
	if nil != err && io.EOF != err {
		return n, err
	}
	self.end += n
```
Increment `end` after error handling

Reply:
The current code defends against the case where `n != 0 and nil != err`.